src/libstrongswan/utils/chunk.h

```diff
@@ -309,6 +309,15 @@ chunk_t chunk_copy_pad(chunk_t dst, chunk_t src, u_char chr);
  */
 int chunk_compare(chunk_t a, chunk_t b);
 
+/**
+ * Compare at most the first N byte of the two chunks
+ */
+static inline bool chunk_equals_prefix(chunk_t a, chunk_t b, size_t len)
```

Comment:
> I'd like to avoid overhead, like making a completely new chunk if a prefix match has to be done, or changing a chunk's length.

What do you mean? Neither has any overhead really.

I'm not entirely opposed to a new function (even if it might not be necessary here), but I really don't like this implementation. It returns TRUE if any bytes at the start of two chunks are common (even a single one). What we want is match received input data against static prefixes and only a complete match of a prefix should be considered a success, so, as I mentioned, it should work like `strpfx()`.

Reply:
Programming overhead. More lines. More text.

The implementation only returns TRUE if exactly len bytes are the same at the beginning of the two chunks, not if any are. The test cases even prove that.

Reply:
> > What do you mean? Neither has any overhead really.
> 
> I don't really see how adding a few lines of code (to override the VID chunk length in the prefix case) is more overhead than adding a new helper function that requires writing unit tests :)

It'd mean first making the chunk and then changing its length, which because you can't do that in one full swoop means declaring one first and then editing it.

> 
> > The implementation only returns TRUE if exactly len bytes are the same at the beginning of the two chunks, not if any are.
> 
> No, the number of bytes compared is controlled by the shortest chunk (or the redundant, as you always set it to a passed chunk's length, additional length).

Well yeah, because otherwise you'd access invalid memory. The len parameter is so you can specify your own length. I just use it like that in the tests because it's the cases I like to test. the vendor-ids branch has an additional test for testing less than chunk.len bytes.

> 
> > The test cases even prove that.
> 
> `chunk_equals_prefix(chunk_a, chunk_b, chunk_a.len)` and `chunk_equals_prefix(chunk_a, chunk_b, chunk_b.len)` both match, so does `chunk_equals_prefix(chunk_a, chunk_b, 512)`, `chunk_equals_prefix(chunk_a, chunk_b, 1)` or even `chunk_equals_prefix(chunk_a, chunk_b, 0)`.
> 
> I think it should be something like this:
> 
> ```c
> static inline bool chunk_equals_prefix(chunk_t a, chunk_t prefix)
> {
> 	return a.ptr != NULL && prefix.ptr != NULL &&
> 		   a.len >= prefix.len && memeq(a.ptr, prefix.ptr, prefix.len);
> }
> ```
> 
> But as you can see it's basically the same as setting `vid_data.len = min(vid_data.len, known_vid.len)` and calling `chunk_equals()` on those chunks.

Yes. The function is just there so you can reuse testing prefixes without having to do it all again or introducing the helper functions later, to avoid code duplication.

Reply:
> It'd mean first making the chunk and then changing its length, which because you can't do that in one full swoop means declaring one first and then editing it.

I really don't see the problem with that and you'd actually change the length of an already declared chunk (only for the IKEv2 case will it be easier to add an additional variable for the known VID, but "creating" that chunk happens anyway, "storing" it in a variable doesn't add overhead).

> Well yeah, because otherwise you'd access invalid memory.

Sure but it makes the function do something that makes not much sense to me for this use case.

> The len parameter is so you can specify your own length.

As it's not enforced that chunks are at least that long this again doesn't make sense to me.

Anyway, please see #393 for further discussions.